### PR TITLE
Use C# 7 deconstruction

### PR DESCRIPTION
**Pull request by drewnoakes**
Changes proposed in this pull request:
 - Add support for C# deconstruction for `KeyValuePair<K,V>`, `IGrouping<K,E>` and `Tuple<T...>`
 - Use deconstruction where possible
 
C#'s deconstruction syntax is pretty nice and can make certain coding patterns much clearer. The primary benefit is that values get more meaningful names than `Key`/`Value`/`Item1`/etc.

The code ends up being a bit shorter and a bit narrower too.

Future work could introduce `ValueTuple<T...>` in place of `Tuple<T...>` and `KeyValuePair<K,V>` to add such names to signatures.

Biggest change was to the impact graph. It used to do a _lot_ of array lookups in `UpdatePathsAndLabels`. Animation seems a bit snappier now, but I didn't measure quantitatively. Code's easier to read now.

What did I do to test the code and ensure quality:
 - Manual review
 - Unit tests
 - Manual testing of all features touched

Has been tested on (remove any that don't apply):
 - GIT 2.10 and above
 - Windows 7 and above

### CODE REVIEW

File: GitCommands/Config/ConfigFile.cs [hunk 1]
```diff
@@ -88,22 +88,22 @@ public string GetAsString()
 
             foreach (var section in ConfigSections)
             {
-                var keys = section.AsDictionary();
+                var dic = section.AsDictionary();
```
Comment (drewnoakes) [inline GitCommands/Config/ConfigFile.cs]: Unrelated change, because code ended up reading strangely without this.

File: GitCommands/Statistics/ImpactLoader.cs [hunk 2]
```diff
@@ -225,24 +225,23 @@ private void LoadModuleInfo(string command, IGitModule module, CancellationToken
         public static void AddIntermediateEmptyWeeks(
             ref SortedDictionary<DateTime, Dictionary<string, DataPoint>> impact, Dictionary<string, DataPoint> authors)
         {
-            foreach (var authorData in authors)
+            foreach (var (author, _) in authors)
```
Comment (drewnoakes) [inline GitCommands/Statistics/ImpactLoader.cs]: Example of ignoring a value during deconstruction via `_`.

File: GitCommands/System.cs [hunk 3]
```diff
@@ -245,4 +245,52 @@ public static string AsForce(this bool force)
             return force ? " -f " : string.Empty;
         }
     }
+
+    public static class DeconstructionExtensions
```
Comment (drewnoakes) [inline GitCommands/System.cs]: .NET Core 2 supports deconstruction of (some of?) these types. For netfx, we must add extensions.

File: GitUI/CommandsDialogs/FormDeleteRemoteBranch.cs [hunk 4]
```diff
@@ -63,13 +63,11 @@ private void OkClick(object sender, EventArgs e)
...
 
-                foreach (var remoteGroup in selectedBranches.GroupBy(b => b.Remote))
+                foreach (var (remote, branches) in selectedBranches.GroupBy(b => b.Remote))
                 {
-                    string remote = remoteGroup.Key;
-
                     EnsurePageant(remote);
 
-                    var cmd = new GitDeleteRemoteBranchesCmd(remote, remoteGroup);
+                    var cmd = new GitDeleteRemoteBranchesCmd(remote, branches);
```
Comment (drewnoakes) [inline GitUI/CommandsDialogs/FormDeleteRemoteBranch.cs]: This looks a bit different to the other cases where `.Key` (etc) was removed. That's because when destructuring `IGrouping<TKey,TElement>` the second value is the grouping itself, re-cast as `IEnumerable<TElement>`.

File: Plugins/Statistics/GitImpact/ImpactControl.cs [hunk 5]
```diff
@@ -393,71 +394,84 @@ private void UpdatePathsAndLabels()
...
-                        author_points.Value[i] = Tuple.Create(
-                            new Rectangle(author_points.Value[i].Item1.Left, (int)(author_points.Value[i].Item1.Top * height_factor),
-                                          author_points.Value[i].Item1.Width, Math.Max(1, (int)(author_points.Value[i].Item1.Height * height_factor))),
-                            author_points.Value[i].Item2);
+                        var (unscaledRect, num) = points[i];
+
+                        var rect = new Rectangle(unscaledRect.Left, (int)(unscaledRect.Top * height_factor),
+                            unscaledRect.Width, Math.Max(1, (int)(unscaledRect.Height * height_factor)));
+
+                        points[i] = Tuple.Create(rect, num);
```
Comment (drewnoakes) [inline Plugins/Statistics/GitImpact/ImpactControl.cs]: `points` elements are reassigned here, which becomes more obvious when there aren't so many array accesses everywhere.

File: TranslationApp/TranslationHelpers.cs [hunk 6]
```diff
@@ -19,12 +19,12 @@ internal static class TranslationHelpers
                 GitCommands.AppSettings.CurrentTranslation = "";
 
                 var translatableTypes = TranslationUtl.GetTranslatableTypes();
-                foreach (var types in translatableTypes)
+                foreach (var (key, types) in translatableTypes)
```
Comment (drewnoakes) [inline TranslationApp/TranslationHelpers.cs]: In this file I left quite a few `key` symbols because I wasn't confident what they should be called. I think they may be `language`.

File: GitCommands/System.cs (hunk 3 above)
Comment (RussKie) [reply GitCommands/System.cs]: Please move this class to its own file

File: Plugins/Statistics/GitImpact/ImpactControl.cs (hunk 5 above)
Comment (gerhardol) [reply Plugins/Statistics/GitImpact/ImpactControl.cs]: I am still not convinced this is the same or that this code works at all...

File: GitCommands/System.cs (hunk 3 above)
Comment (drewnoakes) [reply GitCommands/System.cs]: Done. `System.cs` already had two classes in there, so now there are three separate files.

This file had other horrors too, like the `System` namespace. GE namespaces could do with a review one day.

File: Plugins/Statistics/GitImpact/ImpactControl.cs (hunk 5 above)
Comment (drewnoakes) [reply Plugins/Statistics/GitImpact/ImpactControl.cs]: One good thing with rendering algorithms like this is that bugs usually manifest very visually.
